core/pkg/ingress/controller/controller.go

```diff
@@ -0,0 +1,1076 @@
+			}
+		},
+	}
+
+	eventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ic.syncQueue.Enqueue(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			ic.syncQueue.Enqueue(obj)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			if !reflect.DeepEqual(old, cur) {
+				ic.syncQueue.Enqueue(cur)
+			}
+		},
+	}
+
+	mapEventHandler := cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old, cur interface{}) {
+			if !reflect.DeepEqual(old, cur) {
+				upCmap := cur.(*api.ConfigMap)
+				mapKey := fmt.Sprintf("%s/%s", upCmap.Namespace, upCmap.Name)
+				// updates to configuration configmaps can trigger an update
+				if mapKey == ic.cfg.ConfigMapName || mapKey == ic.cfg.TCPConfigMapName || mapKey == ic.cfg.UDPConfigMapName {
+					ic.recorder.Eventf(upCmap, api.EventTypeNormal, "UPDATE", fmt.Sprintf("ConfigMap %v", mapKey))
+					ic.syncQueue.Enqueue(cur)
+				}
+			}
+		},
+	}
+
+	ic.ingLister.Store, ic.ingController = cache.NewInformer(
+		cache.NewListWatchFromClient(ic.cfg.Client.Extensions().RESTClient(), "ingresses", ic.cfg.Namespace, fields.Everything()),
+		&extensions.Ingress{}, ic.cfg.ResyncPeriod, ingEventHandler)
+
+	ic.endpLister.Store, ic.endpController = cache.NewInformer(
+		cache.NewListWatchFromClient(ic.cfg.Client.Core().RESTClient(), "endpoints", ic.cfg.Namespace, fields.Everything()),
+		&api.Endpoints{}, ic.cfg.ResyncPeriod, eventHandler)
+
+	ic.secrLister.Store, ic.secrController = cache.NewInformer(
+		cache.NewListWatchFromClient(ic.cfg.Client.Core().RESTClient(), "secrets", ic.cfg.Namespace, fields.Everything()),
+		&api.Secret{}, ic.cfg.ResyncPeriod, secrEventHandler)
+
+	ic.mapLister.Store, ic.mapController = cache.NewInformer(
+		cache.NewListWatchFromClient(ic.cfg.Client.Core().RESTClient(), "configmaps", ic.cfg.Namespace, fields.Everything()),
+		&api.ConfigMap{}, ic.cfg.ResyncPeriod, mapEventHandler)
+
+	ic.svcLister.Indexer, ic.svcController = cache.NewIndexerInformer(
+		cache.NewListWatchFromClient(ic.cfg.Client.Core().RESTClient(), "services", ic.cfg.Namespace, fields.Everything()),
+		&api.Service{},
+		ic.cfg.ResyncPeriod,
+		cache.ResourceEventHandlerFuncs{},
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+
+	ic.syncStatus = status.NewStatusSyncer(status.Config{
+		Client:         config.Client,
+		PublishService: ic.cfg.PublishService,
+		IngressLister:  ic.ingLister,
+	})
+
+	return ic
+}
+
+func (ic *GenericController) controllersInSync() bool {
+	return ic.ingController.HasSynced() &&
+		ic.svcController.HasSynced() &&
+		ic.endpController.HasSynced() &&
+		ic.secrController.HasSynced() &&
+		ic.mapController.HasSynced()
+}
+
+// Name returns the healthcheck name
+func (ic GenericController) Name() string {
+	return "Ingress Controller"
+}
+
+// Check returns if the nginx healthz endpoint is returning ok (status code 200)
+func (ic GenericController) Check(_ *http.Request) error {
+	res, err := http.Get("http://127.0.0.1:18080/healthz")
```

Comment:
@aledbf this seems to be very specific to the nginx-based ingress controller, yet is in core/pkg, which I assume is supposed to be generic?
Could you please explain why this is so?
I don't mean this to be criticism, I'm new to this, and just trying to understand this codebase. Is this an example implementation - to be overriden elsewhere, or are all ingress controller running an nginx internally, or is this meant to go away in the future, or am I completely misunderstanding what "generic" means in this context?